[LV] Check for vector-to-scalar casts in legalizer

ErikHogeman opened the pull request:
The code makes assumptions later on the operations and their inputs being scalar in the loops that are processed, so we should make sure this is the case in the legalizer.

ErikHogeman committed:
[LoopVectorize] Check for vector-to-scalar casts in legalizer

The code makes assumptions later on the operations and
their inputs being scalar in the loops that are processed,
so we should make sure this is the case in the legalizer.

artagnon commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,32 @@
+; RUN: opt --force-widen-divrem-via-safe-divisor=false -passes=loop-vectorize -S --debug-only=loop-vectorize < %s 2>&1 | FileCheck %s
```
Missing REQUIRES: asserts, since we're checking debug strings.

artagnon replied:
Would be good to wrap this line like:

```
RUN: opt ... \
RUN:   ... | Filecheck %s
```

artagnon replied:
Is the `--force-widen-divrem-via-safe-divisor=false` needed?

artagnon replied:
> I was able to reproduce it without this flag on some targets, but it seemed more robust to just force the code path that triggers the assert rather than depend on the target specific heuristics which I assume could be changed at any time.

Sounds good. +1 for making it target-independent.

artagnon replied:
Actually, instead of checking debug strings, you can simply check the output (and it will be evident that the example is not vectorized), which can be generated using update_test_checks.py, and add a comment saying that the test was triggering an assertion earlier. That way, we don't have to rely on assertion builds (hence, no need for REQUIRES: asserts).

ErikHogeman replied:
Thanks, I have added such checks now, let me know what you think.

artagnon replied:
Kindly remove `--debug-only=loop-vectorize`, as this will fail in release builds. Also, `2>&1` can be stripped.

ErikHogeman replied:
Ah of course, sorry I missed that, fixed now.

artagnon commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,32 @@
+; CHECK-NOT: Assertion
+
+; Function Attrs: willreturn
+define void @__start() #0 {
```
Suggest renaming `__start` to `vector_to_scalar_cast`?

ErikHogeman replied:
Makes sense, will fix.

artagnon commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,32 @@
+  ret void
+}
+
+attributes #0 = { willreturn }
```
Is this needed?

ErikHogeman replied:
Probably not, will try to remove.

artagnon commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,32 @@
+  %s3 = srem i32 0, %s2
+  br label %bb1
+
+bb3:
```
s/bb3/exit/?

ErikHogeman replied:
I can rename the blocks.

ErikHogeman replied:
Renamed, let me know what you think.

artagnon commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,32 @@
+  %vec1 = insertelement <2 x i16> %vec0, i16 0, i64 1
+  br label %bb0
+
+bb0:
```
Might be good to give nice names to bb0, bb1, and bb2 too.

ErikHogeman replied:
Will rename them.

artagnon commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,32 @@
+  %s0 = phi i32 [ %s1, %bb1 ], [ 1, %entry ]
+  br i1 0, label %bb2, label %bb1
+
+bb1:
```
Can bb0 and bb1 be merged? Did you run `simplifycfg`?

ErikHogeman replied:
I will give it a try.

ErikHogeman replied:
Simplified the loop, let me know what you think.

artagnon commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,32 @@
+  br i1 %exitcond, label %bb0, label %bb3
+
+bb2:
+  %s2 = bitcast <2 x i16> %vec1 to i32
```
Might be good to rename s0, s1, s2, and s3 as iv, add, vec.scalar.cast, and srem.

ErikHogeman replied:
Thanks, will rename.

artagnon commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,32 @@
+; RUN: opt --force-widen-divrem-via-safe-divisor=false -passes=loop-vectorize -S --debug-only=loop-vectorize < %s 2>&1 | FileCheck %s
+
+; CHECK: LV: Not vectorizing: Found unvectorizable type   %s2 = bitcast <2 x i16> %vec1 to i32
+; CHECK-NOT: Assertion
```
Assertion? Is there any debug string that outputs assertion?

ErikHogeman replied:
The test triggers an assert without the code I added in the legalizer, so I just added this to make sure no assert is triggered any longer. Maybe it's not necessary and the test would fail anyway if we trigger an assert? The debug output will print some of these statements before the assert triggers so that's why I was a bit concerned about having only the CHECK.

Should I remove this?

artagnon replied:
Yes, the test would fail anyway. No need to check-not assertion.

ErikHogeman replied:
I have rewritten the checks, let me know what you think.

artagnon commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -943,9 +943,12 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
       // Also, we can't vectorize extractelement instructions.
       if ((!VectorType::isValidElementType(I.getType()) &&
            !I.getType()->isVoidTy()) ||
+          (isa<CastInst>(&I) &&
```
Nit: I don't think the ampersand is required; s/&I/I/.

ErikHogeman replied:
Thanks, will fix.

artagnon reviewed (commented):
Might be good to retitle this s/LoopVectorize/LV/.

ErikHogeman commented:
> Might be good to retitle this s/LoopVectorize/LV/.

Do you mean in the commit message?

Thanks for the review, I will fix your comments tomorrow.

artagnon commented:
> > Might be good to retitle this s/LoopVectorize/LV/.
> 
> Do you mean in the commit message?

You can just re-title the PR using the GitHub interface, and that will be used as the commit message when you merge.

ErikHogeman commented:
> > > Might be good to retitle this s/LoopVectorize/LV/.
> > 
> > 
> > Do you mean in the commit message?
> 
> You can just re-title the PR using the GitHub interface, and that will be used as the commit message when you merge.

Thanks, updated the title.

I also added a use of the "srem" now in the loop, and added a user-defined vectorization width to make it more robust that it triggers the bad code path. The test still crashes without the fix.

artagnon commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,41 @@
+loop:
+  %iv = phi i32 [ %add, %loop ], [ 1, %entry ]
+  %exitcond = icmp ne i32 %iv, 11
+  %add = add nuw nsw i32 %iv, 1
```
Nit: s/%add/%iv.next/.

ErikHogeman replied:
Thanks, fixed.

artagnon commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,41 @@
+}
+
+!0 = !{!0, !1}
+!1 = !{!"llvm.loop.vectorize.width", i32 4}
```
I think you can pass `-force-vector-width=4` to `opt` and remove this.

ErikHogeman replied:
Ah right, I didn't find it for some reason and thought I needed the metadata in the ll-file, this worked fine as well, thanks for the suggestion, fixed.

artagnon commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,41 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt --force-widen-divrem-via-safe-divisor=false -passes=loop-vectorize -S --debug-only=loop-vectorize < %s 2>&1 | FileCheck %s
+
```
Might be good to add a comment like: "The test was crashing earlier due to a vectorization attempt; vector-to-scalar cast is now forbidden in the legalizer, and the test isn't a vectorization candidate now".

ErikHogeman replied:
Good idea, added, thanks!

artagnon reviewed (commented):
LGTM, modulo some comments. Kindly wait for @fhahn or someone else on the reviewer list for the final sign-off.

ErikHogeman commented:
> LGTM, modulo some comments. Kindly wait for @fhahn or someone else on the reviewer list for the final sign-off.

Thanks again for the review!

I will wait until the other reviewers have had time to have a look as well.

fhahn commented on llvm/test/Transforms/LoopVectorize/vector-to-scalar-cast.ll:
```diff
@@ -0,0 +1,40 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; The test was crashing earlier due to a vectorization attempt; vector-to-scalar cast
+; is now forbidden in the legalizer, and the test isn't a vectorization candidate now.
+; RUN: opt -S --force-widen-divrem-via-safe-divisor=false --force-vector-width=4 --passes=loop-vectorize < %s | FileCheck %s
```
```suggestion
; RUN: opt -S -force-widen-divrem-via-safe-divisor=false -force-vector-width=4 --passes=loop-vectorize < %s | FileCheck %s
```

fhahn reviewed (approved):
LGTM, thanks!

ErikHogeman commented:
> LGTM, thanks!

Thank you, I will squash the commits and finalize the patch then.

ErikHogeman commented:
I just realized that while rebasing and squashing I reverted to the old commit message using the full name for LoopVectorize instead of LV as suggested in the review, apologies for that. Let me know if there's anything you want me to do, I assume it's too late to change now.

fhahn commented:
> I just realized that while rebasing and squashing I reverted to the old commit message using the full name for LoopVectorize instead of LV as suggested in the review, apologies for that. Let me know if there's anything you want me to do, I assume it's too late to change now.

Once the PR is created, adjusting the commit messages won't update the PR title/description. Only the PR title/descriptions are used for the final commit. So the commit message of the landed commit should be fine?

ErikHogeman commented:
> Once the PR is created, adjusting the commit messages won't update the PR title/description. Only the PR title/descriptions are used for the final commit. So the commit message of the landed commit should be fine?

Ah, yes you're right, I got confused by the commit in the "commits" tab which had the message I kept while squashing.
OK so no problems then, great, thanks for clarifying that!
